### PR TITLE
ENH: sort=bool keyword argument for index.difference

### LINKED ISSUE #17839: ENH: sort=bool keyword argument for index.difference
I often want to remove certain elements from an index while retaining the original ordering of the index. For this reason a sort keyword in the ``index.difference`` method would be required:

```
In: index = pd.Index([0, 1, 2, 4, 3])
In: index.difference({1, 2})
Out: Int64Index([0, 3, 4], dtype='int64')
# It would be cool to have instead
In: index.difference({1, 2}, sort=False)
Out: Int64Index([0, 4, 3], dtype='int64')
```

I think that this usecase appears frequently and setting the default to ``sort=True`` doesn't affect existing code.

### PR DESCRIPTION
**Pull request by Giftlin**
- [ ] closes #17839
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -2183,7 +2183,7 @@ def _get_consensus_name(self, other):
                 return self._shallow_copy(name=name)
         return self
 
-    def union(self, other):
+    def union(self, other, sort=True):
```
Comment (Licht-T) [inline pandas/core/indexes/base.py]: You need to implement this in subclasses of `Index`, eg. `DatetimeIndex`.
```bash
[pandas] find ./pandas -type f | grep .py$ |  xargs grep -n "def union"                                 17:41:36  ☁  fix-drop_duplicates-when-duplicate-column-names ☂ ✭
./pandas/core/common.py:353:def union(*seqs):
./pandas/core/dtypes/concat.py:214:def union_categoricals(to_union, sort_categories=False, ignore_order=False):
./pandas/core/indexes/base.py:2186:    def union(self, other):
./pandas/core/indexes/datetimes.py:983:    def union(self, other):
./pandas/core/indexes/datetimes.py:1036:    def union_many(self, others):
./pandas/core/indexes/multi.py:2584:    def union(self, other):
./pandas/core/indexes/range.py:416:    def union(self, other):
./pandas/core/indexes/timedeltas.py:496:    def union(self, other):
```

File: pandas/tests/indexes/test_base.py [hunk 2]
```diff
@@ -896,6 +896,33 @@ def test_difference(self):
         assert len(result) == 0
         assert result.name == first.name
 
+    def test_difference_sorting_false(self):
```
Comment (gfyoung) [inline pandas/tests/indexes/test_base.py]: * Reference issue number
* Will need test cases that cause your code to issue the warnings you added

Comment (Giftlin) [reply pandas/tests/indexes/test_base.py]: @gfyoung any changes required?

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (gfyoung) [reply pandas/core/indexes/base.py]: @Giftlin : I don't think you have addressed this comment yet.

Comment (Giftlin) [reply pandas/core/indexes/base.py]: Oh yeah.. will do it 👍

File: pandas/tests/indexes/test_base.py [hunk 3]
```diff
@@ -766,6 +766,103 @@ def test_union(self):
...
+        union = first.union(first, sort=False)
+        assert union is first
+
+        union = first.union([], sort=False)
+        assert union is first
+
+        union = Index([]).union(first, sort=False)
+        assert union is first
+
+        # preserve names
```
Comment (jreback) [inline pandas/tests/indexes/test_base.py]: pls dont' repeat code like this, use parametrize instead.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/indexes/test_base.py (hunk 2 above)
Comment (Licht-T) [reply pandas/tests/indexes/test_base.py]: @Giftlin Some cases will be unsorted even if `sort=True`.
Like this.
https://github.com/pandas-dev/pandas/pull/17878/files#diff-248241e14f081931cc805f9d0e137a5fR2211

**Comment by jreback** [comment]
can you rebase and update according to comments

**Comment by Licht-T** [comment]
@jreback Can I take over this issue?

**Comment by gfyoung** [comment]
@Licht-T : Feel free to write your own patch and submit as a separate PR.

**Pull request closed without merging**